File: rust/2015/day02a/src/main.rs
use std::cmp::min;
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn process_file(filename: &str) -> io::Result<u32> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut total_area: u32 = 0;
    for line_result in reader.lines() {
        let line = line_result?;
        let dimensions = line
            .split('x')
            .map(|x| x.parse().unwrap())
            .collect::<Vec<u32>>();
        let area1 = dimensions[0] * dimensions[1];
        let area2 = dimensions[0] * dimensions[2];
        let area3 = dimensions[1] * dimensions[2];
        let surface_area = (area1 << 1) + (area2 << 1) + (area3 << 1);
        let min_area = min(min(area1, area2), area3);
        total_area += surface_area + min_area
    }
    Ok(total_area)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: zig/2015/day02a/src/main.zig
const std = @import("std");

fn usage(progname: []const u8) void {
    std.debug.print("usage: {s} <input file>\n", .{progname});
    std.process.exit(1);
}

fn min(a: u32, b: u32) u32 {
    return if (a < b) a else b;
}

fn processFile(filename: []const u8) !u32 {
    var file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    var read_buf: [131072]u8 = undefined;
    var reader = file.reader(&read_buf);
    var r_interface = &reader.interface;

    var total_area: u32 = 0;

    while ( true ) {
        const line = r_interface.takeDelimiterExclusive('\n') catch |err| switch (err) {
            error.EndOfStream => break,
            else => return err,
        };
        var parts_iter = std.mem.splitScalar(u8, line, 'x');
        var parts_count: usize = 0;
        var dims: [3]u32 = undefined;
        while ( parts_iter.next() ) |part| {
            if ( parts_count >= dims.len ) return error.InvalidInput;
            const val = std.fmt.parseInt(u32, part, 10) catch return error.InvalidNumber;
            dims[parts_count] = val;
            parts_count += 1;
        }

        const area1 = dims[0] * dims[1];
        const area2 = dims[0] * dims[2];
        const area3 = dims[1] * dims[2];

        const surface_area = area1 * 2 + area2 * 2 + area3 * 2;
        const min_area = min(min(area1, area2), area3);
        
        total_area += surface_area + min_area;
    }

    return total_area;
}

pub fn main() !void {
    var stdout_buffer: [1024]u8 = undefined;
    const allocator = std.heap.page_allocator;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buffer);
    var stdout = &stdout_writer.interface;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if ( args.len < 2 ) {
        usage(args[0]);
    }

    const filename = args[1];

    const result = processFile(filename) catch |err| {
        std.debug.print("error while processing file `{s}': {s}\n", .{filename, @errorName(err)});
        std.process.exit(1);
    };

    try stdout.print("result = {}\n", .{result});
    try stdout.flush();
}
